test/src/unit-cppapi-array.cc

```diff
@@ -396,39 +396,120 @@ TEST_CASE_METHOD(CPPArrayFx, "C++ API: Arrays", "[cppapi][basic]") {
   }
 }
 
-TEST_CASE(
-    "C++ API: Incorrect buffer size and offsets", "[cppapi][invalid-offsets]") {
+TEST_CASE("C++ API: Zero length buffer", "[cppapi][zero-length]") {
   const std::string array_name_1d = "cpp_unit_array_1d";
   Context ctx;
   VFS vfs(ctx);
 
+  tiledb_layout_t write_layout = TILEDB_GLOBAL_ORDER;
+  tiledb_array_type_t array_type = TILEDB_DENSE;
+
+  SECTION("SPARSE") {
+    array_type = TILEDB_SPARSE;
+    SECTION("GLOBAL_ORDER") {
+      write_layout = TILEDB_GLOBAL_ORDER;
+    }
+
+    SECTION("UNORDERED") {
+      write_layout = TILEDB_UNORDERED;
+    }
+  }
+
+  SECTION("DENSE") {
+    array_type = TILEDB_DENSE;
+    SECTION("GLOBAL_ORDER") {
+      write_layout = TILEDB_GLOBAL_ORDER;
+    }
+
+    SECTION("UNORDERED") {
+      write_layout = TILEDB_UNORDERED;
+    }
+  }
+
   if (vfs.is_dir(array_name_1d))
     vfs.remove_dir(array_name_1d);
 
-  ArraySchema schema(ctx, TILEDB_SPARSE);
+  ArraySchema schema(ctx, array_type);
   Domain domain(ctx);
   domain.add_dimension(Dimension::create<int32_t>(ctx, "d", {{0, 1000}}, 1001));
   schema.set_domain(domain);
   schema.add_attribute(Attribute::create<std::vector<int32_t>>(ctx, "a"));
+  schema.add_attribute(Attribute::create<uint64_t>(ctx, "b"));
   Array::create(array_name_1d, schema);
-  Array array(ctx, array_name_1d, TILEDB_WRITE);
 
-  std::vector<int32_t> a, coord = {10, 20, 30};
-  std::vector<uint64_t> a_offset = {0, 0, 0};
-
-  // Test case where backing buffer for vector is non-null, but size is 0.
-  // For bug https://github.com/TileDB-Inc/TileDB/issues/590
   {
+    Array array(ctx, array_name_1d, TILEDB_WRITE);
+
+    std::vector<int32_t> a, coord = {0, 1, 2};
+    std::vector<uint64_t> a_offset = {0, 0, 0};
+    std::vector<uint64_t> b = {1, 2, 3};
+
     a.reserve(10);
     a = {};
```

Comment:
possibly test intermixed empty/occupied instances on single attribute, possibly such as
a = {1, <empty>, <empty>}
a = {<empty>, 2, <empty>}
a = {<empty>, <empty>, 3}
a = {1, <empty>, 3}
?

Reply:
This PR enables var sized attributes where all the elements have zero length. I believe this scenario is outside of the scope.

Reply:
Ok.